persistence/nosql/idgen/impl/src/main/java/org/apache/polaris/ids/impl/SnowflakeIdGeneratorImpl.java

```diff
@@ -0,0 +1,408 @@
+    this.nodeMask = (1L << nodeBits) - 1;
+    this.sequenceBits = sequenceBits;
+    this.sequenceMask = (1L << sequenceBits) - 1;
+    this.epochOffset = epochOffset;
+    this.idGeneratorSource = idGeneratorSource;
+  }
+
+  static void validateArguments(
+      int timestampBits,
+      int sequenceBits,
+      int nodeBits,
+      long epochOffset,
+      IdGeneratorSource idGeneratorSource) {
+    var nowMillis = idGeneratorSource != null ? idGeneratorSource.currentTimeMillis() : -1;
+    var now = Instant.ofEpochMilli(nowMillis);
+    var timestampMax = 1L << timestampBits;
+    checkArgs(
+        () -> checkArgument(idGeneratorSource != null, "IdGeneratorSource must not be null"),
+        () ->
+            checkArgument(
+                nowMillis >= epochOffset,
+                "Clock returns a timestamp %s less than the configured epoch %s",
+                now,
+                Instant.ofEpochMilli(epochOffset)),
+        () ->
+            checkArgument(
+                nowMillis - epochOffset < timestampMax,
+                "Clock already returns a timestamp %s greater of after %s",
+                now,
+                Instant.ofEpochMilli(timestampMax)),
+        () ->
+            checkArgument(
+                nodeBits >= 2
+                    && sequenceBits >= 5
+                    && timestampBits >= 5 // this is REALLY low !
+                    && nodeBits < 64
+                    && sequenceBits < 64
+                    && timestampBits < 64,
+                "value of nodeBits %s or sequenceBits %s or timestampBits %s is too low or too high",
+                nodeBits,
+                sequenceBits,
+                timestampBits),
+        () ->
+            checkArgument(
+                timestampBits + nodeBits + sequenceBits == 63,
+                "Sum of timestampBits + nodeBits + sequenceBits must be == 63"),
+        () -> {
+          if (idGeneratorSource != null) {
+            var nodeId = idGeneratorSource.nodeId();
+            var nodeMax = 1L << nodeBits;
+            checkArgument(
+                nodeId >= 0 && nodeId < nodeMax, "nodeId %s out of range [0..%s[", nodeId, nodeMax);
+          }
+        });
+  }
+
+  static void checkArgs(Runnable... checks) {
+    var violations = new ArrayList<String>();
+    for (Runnable check : checks) {
+      try {
+        check.run();
+      } catch (IllegalArgumentException iae) {
+        violations.add(iae.getMessage());
+      }
+    }
+    if (!violations.isEmpty()) {
+      throw new IllegalArgumentException(String.join(", ", violations));
+    }
+  }
+
+  @Override
+  public long systemIdForNode(int nodeId) {
+    return constructIdUnsafe(timestampMax - 1, 0, nodeId);
+  }
+
+  private long constructIdUnsafe(long timestamp, long sequence, long nodeId) {
+    return (timestamp << timestampShift) | (nodeId << sequenceBits) | sequence;
+  }
+
+  @Override
+  public long constructId(long timestamp, long sequence, long nodeId) {
```

Comment:
This method effectively bypasses the Id generator source. Maybe it shouldn't be public? Having two methods in the same API that generate IDs using different mechanisms seems error-prone. OTOH this method seems only called from `timeUuidToId`, in this same class, so it seems it could be private.